contracts/Protocol.sol

```diff
@@ -20,15 +19,22 @@ contract Protocol is IProtocol, Ownable {
     }
 
     function setProtocolVersion(string calldata deploymentSubset, uint256 protocolVersion, uint256 asOfBlock) external onlyOwner {
-        if (deploymentSubsets[deploymentSubset].version == 0) {
+        if (!deploymentSubsets[deploymentSubset].exists) {
             require(asOfBlock == 0, "initial protocol version must be from block 0");
+            deploymentSubsets[deploymentSubset].currentVersion = protocolVersion;
         } else {
+            uint currentAsOfBlock = deploymentSubsets[deploymentSubset].asOfBlock;
+            uint currentNextVersion = deploymentSubsets[deploymentSubset].nextVersion;
+            if (currentAsOfBlock <= block.number) {
+                deploymentSubsets[deploymentSubset].currentVersion = currentNextVersion;
+            }
+
             require(asOfBlock > block.number, "protocol update can only take place in the future");
-            require(asOfBlock > deploymentSubsets[deploymentSubset].asOfBlock, "protocol upgrade can only take place after the previous protocol update");
-            require(protocolVersion > deploymentSubsets[deploymentSubset].version, "protocol downgrade is not supported");
+            require(asOfBlock > currentAsOfBlock || currentAsOfBlock > block.number, "protocol upgrade can only take place after the previous protocol update, unless previous upgrade is in the future");
```

Comment:
I thought we said we can schedule a new upgrade before the current schedule - as long as the current as of block is after block.number and the protocolVersion is later than currentVersion. I think this line is now redundant and introduces a bug.

what am I missing?

Reply:
Protocol version must be later than current version - this is line 34. We enforce that anyway.
As for scheduling - upgrade must take place in the future (line 32). 
We allow an upgrade to be either after the current planned upgrade (`asOfBlock > currentAsOfBlock`) unless the currently planned upgrade is in the future `currentAsOfBlock > block.number`). This is line 33.
Did I miss anything? let's discuss tomorrow.

Reply:
if `currentAsOfBlock > block.number`:  the require is satisfied.

if the second condition does not hold we get `currentAsOfBlock <= block.number`:  combined with line 32 we get: `asOfBlock > block.number >= currentAsOfBlock` is true, therefore, `asOfBlock > currentAsOfBlock` is true. and so this require is always satisfied.  

so i was wrong, there is no bug, but this line can be removed. what am i missing?

Reply:
You are right, it is redundant.